File: src/main/java/coza/trojanc/receipt/context/impl/SimpleContextDefinition.java
package coza.trojanc.receipt.context.impl;

import coza.trojanc.receipt.context.ContextDefinition;
import coza.trojanc.receipt.context.ContextVariable;

import java.util.Collections;
import java.util.HashMap;
import java.util.Map;

/**
 * A simple implementation of a {@link ContextDefinition}.
 *
 * @author Charl Thiem
 */
public class SimpleContextDefinition implements ContextDefinition {

	/**
	 * Map of fields to resolve
	 */
	private final Map<String, SimpleContextVariable> fields = new HashMap<>();

	/**
	 * Creates a new instance of a {@link SimpleContextDefinition}.
	 */
	public SimpleContextDefinition(){
	}

	/**
	 * Adds a {@link ContextVariable} to the definition.
	 * @param variable The variable to add to the definition.
	 */
	public void addVariable(SimpleContextVariable variable) {
		fields.put(variable.getKey(), variable);
	}

	/**
	 * Gets the map of fields in this {@link ContextDefinition}.
	 * @return The map of fields in this {@link ContextDefinition}.
	 */
	@Override
	public Map<String, ContextVariable> getFields() {
		return Collections.unmodifiableMap(fields);
	}

	/**
	 * Sets the map of fields in this {@link ContextDefinition}.
	 * @param fields The map of fields to set in this  {@link ContextDefinition}.
	 */
	public void setFields(Map<String, SimpleContextVariable> fields) {
		this.fields.clear();
		this.fields.putAll(fields);
	}
}

File: src/main/java/coza/trojanc/receipt/context/impl/SimpleContextDefinition.kt
package coza.trojanc.receipt.context.impl

import coza.trojanc.receipt.context.ContextDefinition
import coza.trojanc.receipt.context.ContextVariable

/**
 * A simple implementation of a [ContextDefinition].
 *
 * @author Charl Thiem
 */
/**
 * Creates a new instance of a [SimpleContextDefinition].
 */
class SimpleContextDefinition : ContextDefinition {

    /**
     * Map of fields to resolve
     */
    val fields = hashMapOf<String, ContextVariable>()

    /**
     * Adds a [ContextVariable] to the definition.
     * @param variable The variable to add to the definition.
     */
    fun addVariable(variable: SimpleContextVariable) {
        fields.put(variable.getKey(), variable)
    }

    /**
     * Gets the map of fields in this [ContextDefinition].
     * @return The map of fields in this [ContextDefinition].
     */
    override fun getFields(): Map<String, ContextVariable> {
        return this.fields; // TODO make a copy
    }

    /**
     * Sets the map of fields in this [ContextDefinition].
     * @param fields The map of fields to set in this  [ContextDefinition].
     */
    fun setFields(fields: Map<String, SimpleContextVariable>) {
        this.fields.clear()
        this.fields.putAll(fields)
    }
}
